accepted/2021/winforms/streamline-application-bootstrap.md

```diff
@@ -0,0 +1,362 @@
+1. Use Roslyn analyzers in Visual Basic scenarios until previous two items are addressed.
+
+
+:thinking: Strategically it could be benefitial to migrate Visual Basic off the Application Framework and *.myapp file in favour of Roslyn source generators and MSBuild properties. This could also significantly reduce efforts in maintaining Visual Studio property pages for Visual Studio projects (e.g. [dotnet/project-system#7236](https://github.com/dotnet/project-system/issues/7236), [dotnet/project-system#7240](https://github.com/dotnet/project-system/issues/7240), and [dotnet/project-system#7241](https://github.com/dotnet/project-system/issues/7241))
+
+
+
+## Stakeholders and Reviewers
+
+<!--
+We noticed that even in the cases where we have specs, we sometimes surprise key
+stakeholders because we didn't pro-actively involve them in the initial reviews
+and early design process.
+
+Please take a moment and add a bullet point list of teams and individuals you
+think should be involved in the design process and ensure they are involved
+(which might mean being tagged on GitHub issues, invited to meetings, or sent
+early drafts).
+-->
+
+* Windows Forms runtime/designer: @dotnet/dotnet-winforms @OliaG @DustinCampbell 
+* Visual Basic: @KathleenDollard  @KlausLoeffelmann
+* Roslyn: @jaredpar @sharwell
+* Project System: @drewnoakes 
+* SDK: dsplaisted 
+* .NET libraries: @ericstj
+* General: @terrajobst @Pilchie 
+
+## Design
+
+<!--
+This section will likely have various subheadings. The structure is completely
+up to you and your engineering team. It doesn't need to be complete; the goal is
+to provide enough information so that the engineering team can build the
+feature.
+
+If you're building an API, you should include the API surface, for example
+assembly names, type names, method signatures etc. If you're building command
+line tools, you likely want to list all commands and options. If you're building
+UI, you likely want to show the screens and intended flow.
+
+In many cases embedding the information here might not be viable because the
+document format isn't text (for instance, because it's an Excel document or in a
+PowerPoint deck). Add links here. Ideally, those documents live next to this
+document.
+-->
+
+
+### 1. Visual Basic 
+
+Ironically Visual Basic apps are in a good position for the designer integration with their Application Framework functionality, i.e. "*.myapp" file:
+
+![image](https://user-images.githubusercontent.com/4403806/114354200-3d9abd80-9bb1-11eb-9d48-a43f299f82e1.png)
+
+The designer surface process should have no issues reading configuration values from a myapp file. Any work further work at this stage pertaining for streamlining the bootstrap is out of scope of this design proposal.
+
+
+
+### 2. C#
+
+```diff
+    static class Program
+    {
+        [STAThread]
+        static void Main()
+        {
++           ApplicationConfiguration.Initialize();
+
+-           Application.EnableVisualStyles();
```

Comment:
Can the designer determine the effective configuration from these existing method calls?

Reply:
This would require the designer compile and parse user code, which will severely impact performance.

Reply:
Would it?  What if it were done with an analyzer/source-generator as part of the compilation that already happens?

Reply:
In theory this generator could be flipped on its head, and have it analyze the source (existing method calls) and output a manifest file for the designer to read. However there is currently no support in Roslyn for source generators that output non-source files, though it has been talked about.

Reply:
It could emit source just the same, attributes if that’s easier for designer to read…. Just an idea if that’s a better user experience. Doesn’t feel too far fetched.

Reply:
There's another reason why we can't rely on code parsing. It won't work in a multi-project setup, where user UI controls reside in one project, and the the `Main` method lives in another project. We can't possibly go and parse all `Main` methods that may exist in a solution, and try to guess which one the designer should base its configuration on.
This proposal makes it possible for the designer to load UI controls in the designer with required settings.

Reply:
Also the designer perf is on a critical path right now, and we need to be extremely mindful of anything we do in a designer that may require additional processing.

Reply:
That's not guaranteed to give correct results. It relies on the user knowing to use `Directory.Build.props` for those settings and not (intentionally or accidentally) overriding them in any individual project. The controls project might even be referenced by multiple exe projects with different settings.

I also don't really think it's an appropriate use of `Directory.Build.props`. These settings have nothing to do with the build at all, and will be adding overhead to evaluation, binlogs, etc.

In the case of controls projects this seems almost like editorconfig in intent 🤷‍♀️

Reply:
> That's not guaranteed to give correct results. It relies on the user knowing to use `Directory.Build.props` for those settings and not (intentionally or accidentally) overriding them in any individual project. 

I don't see this is as a significant problem. Nor being anything significantly different from the current developer experience with SDK-style project. 
AFAIK it is quite common to move shared properties from individual projects to Directory.Build.props.

By default these properties can be configured via the project property page (see: https://github.com/dotnet/project-system/issues/7279), and it will be up to a developer to move them, if necessary.

> The controls project might even be referenced by multiple exe projects with different settings.

The current designer experience is "_we show something which likely doesn't reflect the reality_", and with this proposal it will be "_we do our best to make it WYSIWYG_".
There will be further work required to make it "truly WYSIWYG" and most of this work will be on the VS side, e.g. VS will need to be updated to allow selection of the context, just like it did for the project selection.



> In the case of controls projects this seems almost like editorconfig in intent 🤷‍♀️

If anything with this we'll be sending a wrong message. From  https://editorconfig.org/: "_EditorConfig helps maintain consistent coding styles for multiple developers working on the same project across various editors and IDEs_".